### PR TITLE
Changes to sharing-map needed to merge VSD

### PR DESCRIPTION
**Pull request by martin-cs**
This PR consists of a series of commits taken from the variable-sensitivity-domain branch originally by @hannes-steffenhagen-diffblue and @chrisr-diffblue .  I have separated out the commits (and parts of commits) that just affect the sharing map so these can be reviewed and merged separately.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
Some of these changes might now be redundant and I think it might be best to squash the commits as the messages aren't particularly meaningful now de-contextualised but I figured looking at the code might be the place to start.

**Comment by martin-cs** [comment]
Note that the linter failures are unavoidable.

File: src/util/sharing_map.h [hunk 1]
```diff
@@ -54,6 +54,16 @@ Author: Daniel Poetzl
     typename equalT> \
   type sharing_mapt<keyT, valueT, fail_if_equal, hashT, equalT>
 
+#define SHARING_MAPTV(return_type, V) \
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/sharing_map.h]: Given that this is used in exactly one place I’d like to voice a strong preference for just inlining this macro.

File: src/util/sharing_map.h [hunk 2]
```diff
@@ -413,7 +447,14 @@ class sharing_mapt
   /// - Best case: O(N + H)
   ///
   /// \param [out] view: Empty view
-  void get_view(viewt &view) const;
+  template <class V>
+  void get_view(V &) const;
+  viewt get_view() const
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/sharing_map.h]: I suppose technically this should be a template as well now.

File: src/util/std_expr.h [hunk 3]
```diff
@@ -121,6 +121,18 @@ class symbol_exprt : public nullary_exprt
   }
 };
 
+namespace std
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/std_expr.h]: This is probably harmless but I think it’d nevertheless be preferable if we could find a way to do without this.

(I.e. remove it going forward after VSD is merged, not necessarily here)

**Comment by hannes-steffenhagen-diffblue** [comment]
Linter failures should be addressed with judicious use of `// NOLINT` to explicitly supress linting.

File: src/util/sharing_map.h (hunk 1 above)
Comment (danpoe) [reply src/util/sharing_map.h]: The already existing macro `SHARING_MAPT4` does just the same, so I'd vote for using that one instead.

File: src/util/sharing_map.h [hunk 4]
```diff
@@ -363,6 +386,17 @@ class sharing_mapt
...
+
+  static void insert_view_item(viewt &v, view_itemt &&vi)
+  {
+    v.push_back(vi);
+  }
+
+  static void insert_view_item(sorted_viewt &v, view_itemt &&vi)
+  {
+    v.insert(vi);
+  }
```
Comment (danpoe) [inline src/util/sharing_map.h]: Can we use forwarding references for these two functions? I.e.:
```
template <typename itemT>
static void insert_view_item(viewt &v, itemT &&vi)
{
  v.push_back(std::forward<itemT>(vi));
}
```

File: src/util/sharing_map.h [hunk 5]
```diff
@@ -779,15 +824,15 @@ ::get_sharing_stats_map(Iterator begin, Iterator end)
 }
 #endif
 
-SHARING_MAPT(void)::get_view(viewt &view) const
+SHARING_MAPTV(void, view_type)::get_view(view_type &view) const
```
Comment (danpoe) [inline src/util/sharing_map.h]: Use `SHARING_MAPT4`

File: unit/util/sharing_map.cpp [hunk 6]
```diff
@@ -492,7 +492,7 @@ TEST_CASE("Sharing map views and iteration", "[core][util]")
 
   SECTION("View")
   {
-    typedef std::pair<std::string, std::string> pt;
+    typedef std::pair<irep_idt, std::string> pt;
```
Comment (danpoe) [inline unit/util/sharing_map.cpp]: Suggest changing this back to `std::string`. The intention was to sort the vector in lexicographic order. However, `<` for `irep_idt` sorts according to their sequence number AFAIK.

**Pull request closed without merging**